src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,189 @@
+
+        /// <summary>
+        /// this token is linked to an aggregate token that is passed to the Find References
+        /// operation, this is used solely to trigger the cancellation of an in progress
+        /// find references operation, when the references count hits a given cap.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly SyntaxNode _queriedNode;
+        private readonly ISymbol _queriedSymbol;
+        private readonly ConcurrentSet<ISymbol> _foundDefinitions = new ConcurrentSet<ISymbol>();
+
+        /// <remarks>
+        /// _referencesCount is read and written from multiple threads.
+        /// so all read and write should be through a lock around _gate.
+        /// </remarks>
+        private int _referencesCount;
+
+        /// <remarks>
+        /// If the cap is 0, then there is no cap.
+        /// </remarks>
+        public int SearchCap { get; }
+
+        public CodeLensFindReferencesProgress(
+            ISymbol queriedDefinition,
+            SyntaxNode queriedNode,
+            CancellationTokenSource cancellationTokenSource,
+            int searchCap)
+        {
+            _queriedSymbol = queriedDefinition;
+            _queriedNode = queriedNode;
+            _cancellationTokenSource = cancellationTokenSource;
+
+            SearchCap = searchCap;
+        }
+
+        public bool SearchCapReached
+        {
+            get
+            {
+                if (SearchCap == 0)
+                {
+                    return false;
+                }
+
+                lock (_gate)
+                {
+                    return _referencesCount > SearchCap;
+                }
+            }
+        }
+
+        public int ReferencesCount
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _referencesCount;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns partial symbol locations whose node does not match the given syntaxNode
+        /// </summary>
+        /// <param name="symbol">Symbol whose locations are queried</param>
+        /// <param name="syntaxNode">Syntax node to compare against to exclude location - actual location being queried</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Partial locations</returns>
+        private static IEnumerable<Location> GetPartialLocations(ISymbol symbol, SyntaxNode syntaxNode, CancellationToken cancellationToken)
+        {
+            // Returns nodes from source not equal to actual location
+            IEnumerable<SyntaxNode> syntaxNodes = from syntaxReference in symbol.DeclaringSyntaxReferences
+                                                  let candidateSyntaxNode = syntaxReference.GetSyntax(cancellationToken)
+                                                  where !(syntaxNode.Span == candidateSyntaxNode.Span &&
```

Comment:
no check for null for GetSyntax(,...)? is ISymbol is already checked for soruce type and not generated one?